### PR TITLE
Uplink refund refactor/fixes cyborg spawners not appearing in nuke purchase log

### LINKED ISSUE #3977: Syndicate ending message doesn't include telecrystal cost from syndiborgs
Say you buy a syndiborg and a bunch of other stuff that costs 45 crystals total.
The ending message will show 20 telecrystals spend and not 45.

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
Instead of a snowflake check for cyborg spawners, the uplink now checks vs the uplink datum to see if the item is refundable, and then refunds that cost.

This also fixes syndiborgs not appearing in the purchase log round end. Fixes #3977

Also fixes surplus crates fucking up the TC spent count, which was not a reported bug.

### CODE REVIEW

**Comment by PKPenguin321** [comment]
:+1: Nice refactor, looks easy enough to add new objects to.

File: code/game/objects/items/devices/uplinks.dm [hunk 1]
```diff
@@ -175,14 +175,14 @@ var/list/world_uplinks = list()
...
-			hidden_uplink.uses += S.TC_cost
-			qdel(S)
-			user << "<span class='notice'>Teleporter refunded.</span>"
-		else
-			user << "<span class='warning'>This teleporter is already used!</span>"
+	if(istype(W))
+		for(var/datum/uplink_item/D in subtypesof(/datum/uplink_item))
+			if(D.item == W.type && D.refundable)
+				hidden_uplink.uses += D.cost
+				hidden_uplink.used_TC -= D.cost
```
Comment (duncathan) [inline code/game/objects/items/devices/uplinks.dm]: Change all of these D vars to use initial(). You're looping through paths, not instances.

File: code/game/objects/items/devices/uplinks.dm [hunk 2]
```diff
@@ -175,14 +175,14 @@ var/list/world_uplinks = list()
...
-	if(istype(W, /obj/item/weapon/antag_spawner/nuke_ops))
-		var/obj/item/weapon/antag_spawner/nuke_ops/S = W
-		if(!S.used)
-			hidden_uplink.uses += S.TC_cost
-			qdel(S)
-			user << "<span class='notice'>Teleporter refunded.</span>"
-		else
-			user << "<span class='warning'>This teleporter is already used!</span>"
+	if(istype(W))
+		for(var/datum/uplink_item/D in subtypesof(/datum/uplink_item))
```
Comment (duncathan) [inline code/game/objects/items/devices/uplinks.dm]:
```
for(var/path in subtypesof(/datum/uplink_item))
    var/datum/uplink_item/D = path
```

**Comment by duncathan** [comment]
@KorPhaeron

**Comment by KorPhaeron** [comment]
@duncathan 

Done

**Pull request merged**